arch/arm/src/arm/Toolchain.defs

```diff
@@ -113,15 +113,28 @@ endif
 
 # Default toolchain
 
-CC = $(CROSSDEV)gcc
-CXX = $(CROSSDEV)g++
-CPP = $(CROSSDEV)gcc -E -P -x c
-LD = $(CROSSDEV)ld
-STRIP = $(CROSSDEV)strip --strip-unneeded
-AR = $(CROSSDEV)ar rcs
-NM = $(CROSSDEV)nm
+CC      = $(CROSSDEV)gcc
+CXX     = $(CROSSDEV)g++
+CPP     = $(CROSSDEV)gcc -E -P -x c
+STRIP   = $(CROSSDEV)strip --strip-unneeded
 OBJCOPY = $(CROSSDEV)objcopy
 OBJDUMP = $(CROSSDEV)objdump
+LD      = $(CROSSDEV)ld
+AR      = $(CROSSDEV)ar rcs
+NM      = $(CROSSDEV)nm
+
+# Link Time Optimization
+
+ifeq ($(CONFIG_LTO_FULL),y)
+  MAXOPTIMIZATION += -flto
+  ifeq ($(CONFIG_ARM_TOOLCHAIN),GNU_EABI)
+    MAXOPTIMIZATION += -fuse-linker-plugin
+    MAXOPTIMIZATION += -fno-builtin
+    LD := $(CROSSDEV)gcc
```

Comment:
why link with ld when CONFIG_LTO_FULL==n, but with gcc when CONFIG_LTO_FULL==y? and complicate src/arm/Makefile.

Reply:
flto has no effect on ld. It is a gcc option.

Reply:
I prefer either link with ld or  link with gcc, but not switch back and forth.

Reply:
@yamt do you have any comment?

Reply:
i will take a look. hopefully today.

Reply:
* i agree with @xiaoxiang781216 that it's better to use either ld or cc. at least for a board.
* i'm against LD=gcc. cc and ld take different arguments. little point to "reuse" LD var. if you want to use cc to link, it's simpler to use CC.

Reply:
why this thread was marked resolved?

Reply:
Since gcc and clang has the different requirement:

1. gcc require invoke gcc instead ld to tirgger lto
2. but clang  doesn't require this special hack

Do you have better solution?

Reply:
> Do you have better solution?

* write a wrapper script and use it as LD. something like https://lkml.iu.edu/hypermail/linux/kernel/1402.1/00048.html
* or, some complicated setup like https://github.com/yamt/garbage/blob/baad156d2b19687d40bd603188756fca6275af6e/lto/Makefile#L36-L46
* or, always use CC to link

i feel the first one is the least invasive.

Reply:
The first method is very like this patch, but look more modular. @anchao what your thinking?

Reply:
IMO it is just better always use CC to link in case of GCC and remove override in case of LTO.